lib/linked_list.rb

```diff
@@ -12,152 +12,296 @@ def initialize(value, next_node = nil)
+    return if temp_node == nil
 
-    # method to print all the values in the linked list
-    # Time Complexity:  
-    # Space Complexity
-    def visit
-      raise NotImplementedError
+    if temp_node.data == value
+      @head = temp_node.next
+      return
     end
 
-    # method to delete the first node found with specified value
-    # Time Complexity:  
-    # Space Complexity
-    def delete(value)
-      raise NotImplementedError
+    until temp_node.next == nil
+      temp_next = temp_node.next
+      if temp_next.data == value
+        temp_node.next = temp_next.next
+        return
+      end
+      temp_node = temp_next
     end
+    return
+  end
+
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def reverse
+    current = @head
+    prev = nil
 
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity:  
-    # Space Complexity
-    def reverse
-      raise NotImplementedError
+    while current
+      temp = current.next
+      current.next = prev
+      prev = current
+      current = temp
     end
+    @head = prev
+    return
+  end
+
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def find_middle_value
+    return self.get_at_index(self.length / 2)
+  end
 
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)
+    temp_node = @head
+    return nil if temp_node == nil
+    counter = 0
+    n_away = temp_node
+    until counter == n
+      return nil if n_away.next == nil
+      n_away = n_away.next
+      counter += 1
+    end
 
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity:  
-    # Space Complexity
-    def find_middle_value
-      raise NotImplementedError
+    until n_away.next == nil
+      temp_node = temp_node.next
+      n_away = n_away.next
     end
+    return temp_node.data
+  end
 
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity:  
-    # Space Complexity
-    def find_nth_from_end(n)
-      raise NotImplementedError
+  # checks if the linked list has a cycle. A cycle exists if any node in the
+  # linked list links to a node already visited.
+  # returns true if a cycle is found, false otherwise.
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(n), where n is the number of nodes in the list
+  def has_cycle

```

Comment:
Very clever use of a hash!  You can also do it with O(1) space complexity using [two pointers](https://www.geeksforgeeks.org/detect-loop-in-a-linked-list/) one moving twice as fast as the other.